numpy/core/src/multiarray/mapping.c

```diff
@@ -488,7 +488,11 @@ prepare_index(PyArrayObject *self, PyObject *index,
                 indices[curr_idx].type = HAS_0D_BOOL;
 
                 /* TODO: This can't fail, right? Is there a faster way? */
-                if (PyObject_IsTrue((PyObject *)arr)) {
+                int istrue = PyObject_IsTrue((PyObject *)arr);
+                if (istrue == -1) {
+                    return -1;
```

Comment:
The comment above would be nice to adjust or just remove (yes, there is a faster way as `n = ((npy_bool *)PyArray_BYTES(arr))[0] != 0`, but it doesn't really matter.)

The real thing that needs fixing though: `return -1` here is wrong, it needs to be a `goto` to ensure cleaning up happens.  So unless you want to think a bit about the above that cannot error, we need to fix that.

Reply:
I will update the comment to add above suggestion in future. So rather than `return -1` it should be` goto failed_building_indices` right